Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,320 @@
+* `G.Adj x y` means that there is an edge `e` having `x` and `y` as its ends.
+* `G.IsLoopAt e x` means that `e` is a loop edge with both ends equal to `x`.
+* `G.IsNonloopAt e x` means that `e` is a non-loop edge with one end equal to `x`.
+
+## Implementation notes
+
+Unlike the design of `SimpleGraph`, the vertex and edge sets of `G` are modelled as sets
+`V(G) : Set α` and `E(G) : Set β`, within ambient types, rather than being types themselves.
+This mimics the 'embedded set' design used in `Matroid`, which seems to be more convenient for
+formalizing real-world proofs in combinatorics.
+
+A specific advantage is that this allows subgraphs of `G : Graph α β` to also exist on
+an equal footing with `G` as terms in `Graph α β`,
+and so there is no need for a `Graph.subgraph` type and all the associated
+definitions and canonical coercion maps. The same will go for minors and the various other
+partial orders on multigraphs.
+
+The main tradeoff is that parts of the API will need to care about whether a term
+`x : α` or `e : β` is a 'real' vertex or edge of the graph, rather than something outside
+the vertex or edge set. This is an issue, but is likely amenable to automation.
+
+## Notation
+
+Reflecting written mathematics, we use the compact notations `V(G)` and `E(G)` to
+refer to the `vertexSet` and `edgeSet` of `G : Graph α β`.
+
+-/
+
+variable {α β : Type*} {x y z u v w : α} {e f : β}
+
+open Set
+
+/-- A multigraph with vertex set `vertexSet : Set α` and edge set `edgeSet : Set β`,
+as described by a predicate describing whether an edge `e : β` has ends `x` and `y`.
+For definitional reasons, we include `edgeSet` as a structure field
+even though it can be inferred from `IsLink` via `edge_mem_iff_exists_isLink`;
+see `Graph.mk'` for a constructor that does not use `edgeSet`. -/
+structure Graph (α β : Type*) where
+  /-- The vertex set. -/
+  vertexSet : Set α
+  /-- The edge set. -/
+  edgeSet : Set β
+  /-- The binary incidence predicate, stating that `x` and `y` are the ends of an edge `e`. -/
+  IsLink : β → α → α → Prop
+  /-- If `e` goes from `x` to `y`, it goes from `y` to `x`. -/
+  isLink_symm : ∀ ⦃e⦄, e ∈ edgeSet → (Symmetric <| IsLink e)
+  /-- An edge is incident with at most one pair of vertices. -/
+  eq_or_eq_of_isLink_of_isLink : ∀ ⦃e x y v w⦄, IsLink e x y → IsLink e v w → x = v ∨ x = w
+  /-- An edge `e` is incident to something if and only if `e` is in the edge set. -/
+  edge_mem_iff_exists_isLink : ∀ e, e ∈ edgeSet ↔ ∃ x y, IsLink e x y
+  /-- If some edge `e` is incident to `x`, then `x ∈ V`. -/
+  vx_mem_left_of_isLink : ∀ ⦃e x y⦄, IsLink e x y → x ∈ vertexSet
+
+namespace Graph
+
+variable {G : Graph α β}
+
+/-- `V(G)` denotes the `vertexSet` of a graph `G`. -/
+scoped notation "V(" G ")" => Graph.vertexSet G
+
+/-- `E(G)` denotes the `edgeSet` of a graph `G`. -/
+scoped notation "E(" G ")" => Graph.edgeSet G
+
+/-! ### Edge-vertex-vertex incidence -/
+
+lemma IsLink.edge_mem (h : G.IsLink e x y) : e ∈ E(G) :=
```

Comment:
I see `mem_edgeSet` below. Maybe `IsLink.mem_edgeSet` is more consistent?

I think I'm ok with this though, so long as we document that "if `G.IsLink e x y` is true then we refer to `e` as an `edge` and `x` and `y` as `verts`/`vtxs` in names"

Reply:
What do you think of just `left` and `right` to refer to the vertices when in the `IsLink` namespace?